src/Experimental/isomorphism.jl

```diff
@@ -0,0 +1,881 @@
+        T = Vector{Tuple{eltype(g1), eltype(g2)}}
+        ch::Channel{T} = Channel(ctype=T) do c
+            vf2(make_callback(c), g1, g2, IsomorphismProblem, 
+                           vertex_relation=vertex_relation,
+                           edge_relation=edge_relation)
+        end
+        return ch
+    else
+        throw(ArgumentError("Keyword argument alg must be :vf2"))
+    end
+end
+
+"""
+    VF2State{G, T}
+Structure that is internally used by vf2
+"""
+struct VF2State{G, T}
+    g1::G
+    g2::G
+    core_1::Vector{T}
+    core_2::Vector{T}
+    in_1::Vector{T}
+    in_2::Vector{T}
+    out_1::Vector{T}
+    out_2::Vector{T}
+
+    function VF2State(g1::G, g2::G) where {G <: AbstractSimpleGraph{T}} where {T <: Integer}
+        n1 = nv(g1)
+        n2 = nv(g2)
+        core_1 = zeros(T, n1)
+        core_2 = zeros(T, n2)
+        in_1 = zeros(T, n1)
+        in_2 = zeros(T, n2)
+        out_1 = zeros(T, n1)
+        out_2 = zeros(T, n2)
+
+        return new{G, T}(g1, g2, core_1, core_2, in_1, in_2, out_1, out_2)
+    end
+end
+
+"""
+    vf2(callback, g1, g2, problemtype; vertex_relation=nothing, edge_relation=nothing)
+Iterates over all isomorphism between the graphs `g1` (or subgraphs thereof) and `g2`.
+The problem that is solved depends on the value of `problemtype`:
+- IsomorphismProblem: Only isomorphisms between the whole graph `g1` and `g2` are considered.
+- SubGraphIsomorphismProblem: All isomorphism between subgraphs of `g1` and `g2` are considered.
+- InducedSubGraphIsomorphismProblem: All isomorphism between vertex induced subgraphs of `g1` and `g2` are considered.
+
+Upon finding an isomorphism, the function `callback` is called with a vector `vmap` as an argument.
+`vmap` is a vector where `vmap[v] == u` means that vertex `v` in `g2` is mapped to vertex `u` in `g1`.
+If the algorithm should look for another isomorphism, then this function should return `true`.
+
+### Optional Arguments
+- `vertex_relation`: A binary function that takes a vertex from `g1` and one from `g2`. An
+    isomorphism only exists if this function returns `true` for all matched vertices.
+- `edge_relation`: A binary function that takes an edge from `g1` and one from `g2`. An
+    isomorphism only exists if this function returns `true` for all matched edges.
+### References
+Luigi P. Cordella, Pasquale Foggia, Carlo Sansone, Mario Vento
+“A (Sub)Graph Isomorphism Algorithm for Matching Large Graphs”
+"""
+function vf2(callback::Function, g1::G, g2::G, problemtype::GraphMorphismProblemType; 
+             vertex_relation::Union{Nothing, Function}=nothing, 
+             edge_relation::Union{Nothing, Function}=nothing) where {G <: AbstractSimpleGraph}
+    if has_self_loops(g1) || has_self_loops(g2)
+        throw(ArgumentError("vf2 does not support self-loops at the moment"))
+    end
+
+    if nv(g1) < nv(g2) || (problemtype == IsomorphismProblem && nv(g1) != nv(g2))
+        return 
```

Comment:
Does this empty `return` cause the function to be type-unstable? If `vf2match!` doesn't return anything, then `vf2` should probably be renamed `vf2!`, right?

Reply:
`vf2` doesn't mutate any input arguments and never returns anything. The only thing it does when finding an isomorphism is calling the function `callback` and it is up to the user to decide if he wants to mutate something then.

This is a bit complicated and the reason why I made all the other functions such as `has_iso`.

Reply:
`vf2match!` returns a boolean, and `vf2` will return `vf2match!` since it's the last thing evaluated on line 522. If you want to ignore the return value of `vf2match!`, you need an explicit `return` after that line.

Reply:
You are right, I always forget that Julia workings differently from other languages here.